Support for multiclass, word embeddings, configuration file and new datasets

cahya-wirawan opened the pull request:
Hi,

I added following functionalities:
- multiclass classification 
- pre-trained word embedding using word2vec and GloVe
- configuration file in yaml format
- new dataset 20newsgroup (loaded using sklearn.datasets)
- loading multiclass text based dataset from local directory

And also path to the movie rating dataset has been moved to the configuration file. Thanks.

cahya-wirawan committed:
Following functionalities have been added:
- multiclass functionality
- 20 newsgroup loader
- yaml configuration file
- using pre trained word2vec and GloVe embedded words

cahya-wirawan committed:
Fixed issues with word2vec loader

cahya-wirawan committed:
Added the possibility to disable the word embeddings by using the option enable_word_embeddings
or set word_embeddings.default to an empty string in config.yml

cahya-wirawan committed:
- API changes for TF 1.0
- Added a functionality to load text datasets from local files

cahya-wirawan committed:
Updated info about the parameter of get_datasets_localdata

cahya-wirawan committed:
Merge remote-tracking branch 'upstream/master'

cahya-wirawan committed:
Added enable_word_embeddings in README file

cahya-wirawan committed:
- added softmax function
- calculate the probability from the score using
  softmax function and save it in prediction.csv file
- fixed an issue with empty datasets["target_names"]

CMWENLIU commented:
Hi @cahya-wirawan 
Thank you so much for the functionality of multiclass classification you did.
I still have issues when loading my own local data, after following I did:

1, saved text files with categories as subfolder names in the folder: /data/bbcdata
   and there are 5 folders with corresponding txt files in bbcdata folder: "business","entertainment","politics","sport","tech"
2, updated the config.yml file as following
```
line 16: default: localdata
line 52: container_path: "/data/bbcdata"
```
Did I missing something to run the ./train.py
Could you help me about that?
Thank you so much!

Aven

cahya-wirawan committed:
Added a dynamic learning rate with high value at the beginning to get
higher accuracy faster. The learning rate decay exponentially
from 0.003 to 0.0001.

cahya-wirawan committed:
Fixed a comment

cahya-wirawan committed:
Added a new option decay_coefficient

amatosc commented on data_helpers.py:
```diff
@@ -63,3 +43,129 @@ def batch_iter(data, batch_size, num_epochs, shuffle=True):
+    return [x_text, y]
+
+
+def load_embedding_vectors_word2vec(vocabulary, filename, binary):
```
Hey! this change is pretty amazing, I would like to know if I could use this in instance of this function:
``` py
 >>> from gensim.models.keyedvectors import KeyedVectors
 >>> word_vectors = KeyedVectors.load_word2vec_format('/tmp/vectors.txt', binary=False)  # C text format
 >>> word_vectors = KeyedVectors.load_word2vec_format('/tmp/vectors.bin', binary=True)  # C binary format
```

usmaann commented:
How much is the expected training time ? and how many steps are needed to get good accuracy results/???

usmaann commented:
> Hi @cahya-wirawan
> Thank you so much for the functionality of multiclass classification you did.
> I still have issues when loading my own local data, after following I did:
> 
> 1, saved text files with categories as subfolder names in the folder: /data/bbcdata
> and there are 5 folders with corresponding txt files in bbcdata folder: "business","entertainment","politics","sport","tech"
> 2, updated the config.yml file as following
> 
> ```
> line 16: default: localdata
> line 52: container_path: "/data/bbcdata"
> ```
> Did I missing something to run the ./train.py
> Could you help me about that?
> Thank you so much!
> 
> Aven
> 
> @cahya-wirawan
> Following is the error I get using local data for multi-class data:
> Could you help me about this?
> Thanks a lot!
> 
> ```
> Loading data...
> Traceback (most recent call last):
>   File "./train.py", line 72, in <module>
>     random_state=cfg["datasets"][dataset_name]["random_state"])
>   File "/home/xxliu10/repo/cahya_cnn/cnn-text-classification-tf/data_helpers.py", line 93, in get_datasets_localdata
>     random_state=random_state)
>   File "/home/xxliu10/anaconda3/lib/python3.6/site-packages/sklearn/datasets/base.py", line 232, in load_files
>     data = [d.decode(encoding, decode_error) for d in data]
>   File "/home/xxliu10/anaconda3/lib/python3.6/site-packages/sklearn/datasets/base.py", line 232, in <listcomp>
>     data = [d.decode(encoding, decode_error) for d in data]
> UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa3 in position 257: invalid start byte
> ```

Hi are you able to fix this issue? I am facing the same issue

usmaann commented:
> > Hi @cahya-wirawan
> > Thank you so much for the functionality of multiclass classification you did.
> > I still have issues when loading my own local data, after following I did:
> > 1, saved text files with categories as subfolder names in the folder: /data/bbcdata
> > and there are 5 folders with corresponding txt files in bbcdata folder: "business","entertainment","politics","sport","tech"
> > 2, updated the config.yml file as following
> > ```
> > line 16: default: localdata
> > line 52: container_path: "/data/bbcdata"
> > ```
> > Did I missing something to run the ./train.py
> > Could you help me about that?
> > Thank you so much!
> > Aven
> > @cahya-wirawan
> > Following is the error I get using local data for multi-class data:
> > Could you help me about this?
> > Thanks a lot!
> > ```
> > Loading data...
> > Traceback (most recent call last):
> >   File "./train.py", line 72, in <module>
> >     random_state=cfg["datasets"][dataset_name]["random_state"])
> >   File "/home/xxliu10/repo/cahya_cnn/cnn-text-classification-tf/data_helpers.py", line 93, in get_datasets_localdata
> >     random_state=random_state)
> >   File "/home/xxliu10/anaconda3/lib/python3.6/site-packages/sklearn/datasets/base.py", line 232, in load_files
> >     data = [d.decode(encoding, decode_error) for d in data]
> >   File "/home/xxliu10/anaconda3/lib/python3.6/site-packages/sklearn/datasets/base.py", line 232, in <listcomp>
> >     data = [d.decode(encoding, decode_error) for d in data]
> > UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa3 in position 257: invalid start byte
> > ```
> 
> Hi are you able to fix this issue? I am facing the same issue

Anybody can give the solution of this problem?

![image](https://user-images.githubusercontent.com/38275359/48683022-339e4b80-ebff-11e8-8f37-ad3aa14f12f0.png)
